### ISSUE TITLE
mssqlnative: Query not returning ID

### ISSUE THREAD

**Issue by jorgemfm** [ADOdb/ADOdb#185]
Version 5.20.2
mssqlnative Driver

If a textfield in inserted empty, the query does not return the ID:

Ex:
$data['field1'] =1;
$data['field2'] = '';
$conn->AutoExecute($table, $data, 'INSERT');
$conn->Insert_ID(); //does not return ID

$data['field1'] =1;
$data['field2'] = 'string of text';
$conn->AutoExecute($table, $data, 'INSERT');
$conn->Insert_ID(); // returns ID

The problem lies in preg_match on the _query method, but i'm not very good at regex so I was unable to solve it.

On version 5.19 works OK but this versions has this problem: failure on Insert_ID() if the insert statement contains a semicolon in a value string

**Comment by mnewnham** [comment]
based on your above example, do you mean:

``if the insert statement contains a null in a value string``

or should your example 1 say

``$data['field2'] = 'some text with a ; in it';``

**Comment by jorgemfm** [comment]
No, it's really like I said, if I pass an empty string like this: 

$data['field2'] = ''; 

The ID is never returns because its never entrer the condition:

**Comment by jorgemfm** [comment]
No, it's really like I said, if I pass an empty string like this: 

$data['field2'] = '';

The ID is never returned because it never enters the condition:

if(preg_match('......'){
....
}

on the _query() method

**Comment by dregad** [comment]
See #96 for the original issue introducing the new regex. Looking at test code provided by @stekule, it appears the empty string was not considered; I believe the simplest fix would be to add an alternative to match the empty string, i.e. 

```
/^\W*insert\s(?:\'\'|(?:(?:\'\')*\'[^\']+\'(?:\'\')*)|[^;\'])*;?$/i
                ^^^^^ new bit
```

The current regex is not very efficient, as it goes through the matching group for every single char after the initial *insert*. and needs over 400 steps [1] to match a simple statement like `INSERT INTO schema.my_table ( c1, c2, c3 ) VALUES ( '', 'x;', 12 )`.

Leveraging the insert statement's syntax, we can modify the regex to only parse the values. Also, moving the complex group after simpler ones is better. Something like this reduces the number of steps to 72 [1] for the same SQL statement:
```
/^\W*insert[\s\w()",.]+values\s*\((?:[^;\']|\'\'|(?:(?:\'\')*\'[^\']+\'(?:\'\')*))*;?$/i
```

- *[\s\w()",.]+values\s*\(* should match `[ INTO ] { <object> } [ ( column_list ) ]` until VALUES keyword and following opening parenthesis is found
- move *[^;\']* to beginning of matching group

I can't test on MSSQL, so I let you check whether this updated regex works for you.

[1] steps calculation per https://regex101.com/r/yU0gD9/4 regex debugger

**Comment by mnewnham** [comment]
On a more general side, I'm intrigued to know why our regexp needs to validate a well-formed insert statement. I get that we only want to attach the ``select scope(identity)``  onto an insert, but beyond validation that the statement is ``insert into``, what other issue are we looking at?

**Comment by mnewnham** [comment]
I've just done a simple ADOdb version comparison, bases on insertion of 1000 records into a small table.  The last version of ADOdb that did not automatically retrieve the insert id was 5.18a.  
 1. attaching that driver results in a time of 5.5 seconds for 1000 records. 
 2. Replacing that driver with 5.19 and above results in an insertion time of 14.2  seconds
 3. Remarking out the test that causes the ``identitySql`` to be attached to the query in versions 5.19-5.20.2 returns the insert time to 5.5 seconds

Now if we add back into the function _insertid() the code

````php
function _insertid()
{
	$rez = sqlsrv_query($this->_connectionID,$this->identitySQL);
	sqlsrv_fetch($rez);
	$this->lastInsertID = sqlsrv_get_field($rez, 0);
}
````

This only increases the insert time from 5.5 seconds to 6.8 seconds. In addition, inserts that do not require the insert id to be retrieved get the speed boost.

Clearly, this is a trivial example, but I believe there needs to be a serious discussion about the current methodology. Anyone who can confirm or disprove these figures above, please do

**Comment by dregad** [comment]
> but beyond validation that the statement is insert into, what other issue are we looking at?

No idea. I'm not even using this driver and did not bother to look at the big picture.

**Comment by jorgemfm** [comment]
Did a few testes and both regex proposed by @dregad seems to work OK.

Thank you

**Comment by dregad** [comment]
Thanks for testing @jorgemfm.

@mnewnham let me know how would you like to proceed with this. I would propose to patch the regex as a 5.20.x hotfix, and address the deeper issue you identified as a separate issue.

**Comment by dregad** [comment]
I just realized that I never actually applied the modified regex as a hotfix for the 5.20 branch, as mentioned in https://github.com/ADOdb/ADOdb/issues/185#issuecomment-169592672. This will be included in upcoming 5.20.8.
